src/VBORenderer.cc

```diff
@@ -0,0 +1,805 @@
+		if (OpenSCAD::debug != "") PRINTDB("create_vertex point_c : [%d, %d, %d]",
+			pc[0] % pc[1] % (pc[2]));
+
+	} else {
+		if (OpenSCAD::debug != "") PRINTDB("create_vertex bad points size = %d", points.size());
+	}
+}
+
+void VBORenderer::create_triangle(VertexArray &vertex_array, const Color4f &color,
+				const Vector3d &p0, const Vector3d &p1, const Vector3d &p2,
+				size_t primitive_index,
+				double z_offset, size_t shape_size,
+				size_t shape_dimensions, bool outlines,
+				bool mirror) const
+{
+	double ax = p1[0] - p0[0], bx = p1[0] - p2[0];
+	double ay = p1[1] - p0[1], by = p1[1] - p2[1];
+	double az = p1[2] - p0[2], bz = p1[2] - p2[2];
+	double nx = ay*bz - az*by;
+	double ny = az*bx - ax*bz;
+	double nz = ax*by - ay*bx;
+	double nl = sqrt(nx*nx + ny*ny + nz*nz);
+	
+	if (!vertex_array.data()) return;
+	
+	shared_ptr<VertexData> vertex_data = vertex_array.data();
+
+	if (shader_write_index)
+		add_shader_attributes(vertex_array, color, {p0, p1, p2}, 0,
+		      			primitive_index, z_offset, shape_size,
+					shape_dimensions, outlines, mirror);
+	addAttributeValues(*(vertex_data->positionData()), p0[0], p0[1], p0[2]);
+	if (!mirror) {
+		if (shader_write_index)
+			add_shader_attributes(vertex_array, color, {p0, p1, p2}, 1,
+						primitive_index, z_offset, shape_size,
+						shape_dimensions, outlines, mirror);
+		addAttributeValues(*(vertex_data->positionData()), p1[0], p1[1], p1[2]);
+	}
+	if (shader_write_index)
+		add_shader_attributes(vertex_array, color, {p0, p1, p2}, 2,
+					primitive_index, z_offset, shape_size,
+					shape_dimensions, outlines, mirror);
+	addAttributeValues(*(vertex_data->positionData()), p2[0], p2[1], p2[2]);
+	if (mirror) {
+		if (shader_write_index)
+			add_shader_attributes(vertex_array, color, {p0, p1, p2}, 1,
+						primitive_index, z_offset, shape_size,
+						shape_dimensions, outlines, mirror);
+		addAttributeValues(*(vertex_data->positionData()), p1[0], p1[1], p1[2]);
+	}
+	if (vertex_data->hasNormalData()) {
+		addAttributeValues(3, *(vertex_data->normalData()), (nx/nl), (ny/nl), (nz/nl));		
+	}
+	if (vertex_data->hasColorData()) {
+		addAttributeValues(3, *(vertex_data->colorData()), color[0], color[1], color[2], color[3]);
+	}
+}
+
+void VBORenderer::create_surface(shared_ptr<const Geometry> geom, VertexArray &vertex_array,
+				 csgmode_e csgmode, const Transform3d &m, const Color4f &color) const
+{
+	shared_ptr<const PolySet> ps = dynamic_pointer_cast<const PolySet>(geom);
+	shared_ptr<VertexData> vertex_data = vertex_array.data();
+
+	if (!ps || !vertex_data) { return; }
+
+	bool mirrored = m.matrix().determinant() < 0;
+	size_t triangle_count = 0;
+
+	if (ps->getDimension() == 2) {
+		create_polygons(geom, vertex_array, csgmode, m, color);
+	} else if (ps->getDimension() == 3) {
+		VertexStates &vertex_states = vertex_array.states();
+		size_t last_size = vertex_data->sizeInBytes();
+
+		for (const auto &poly : ps->polygons) {
+			Vector3d p0 = m * poly.at(0);
```

Comment:
From my profiling I noticed that a most of the preview compute time (57% !) is spent in this `create_surface` function, particularly applying these Eigen matrices.

Again builds with optimization are drastically faster than unoptimized in this case, but I think the previous implementation left the matrix transformations for the GPU to handle.  Is there a reason to do it on CPU here?

Reply:
Yes. OpenCSG performance is increased by doing the multiplications once to the objects.

The original profiling done showed that OpenCSG was slowed down by the glMultMatrix calls in render, and it appears most drivers implement those calls on the CPU anyway. There is no option to multiply once in the old renderer code since each call is rendering the geometry every time the display is updated.

As to GPU use, which would require shaders, OpenCSG cannot use shaders in it's draw loop. There are internal shaders used during the render call, and these conflict with current OpenGL shaders.

Until a new library is used for CSG, shaders can only be used on the final pass draws after OpenCSG has done it's work.

Reply:
OK, thanks for the reply.  I'll leave it as is then.